Support binding to form and form file parameters in minimal actions

martincostello opened the pull request:
**PR Title**

Support binding to the form and files in the form from parameters in minimal actions.

**PR Description**

Add basic support for binding to parameters of types ~`IFormCollection`~,  `IFormFileCollection`, and `IFormFile` on the `HttpRequest` in minimal actions.

I've just done this as a draft for now as while the new and existing `RequestDelegateFactory` tests pass, I'm sure there's better ways to deal with some of the reflection/expression binding stuff I've put in.

## TODO

- [x] ~A public `IFromFileMetadata` implementation, such as a metadata class/attribute.~
- [x] ~API review for interface and the above missing implementation of it. https://github.com/dotnet/aspnetcore/issues/35175~
- [x] ~More test cases probably for more permutations of actions using these parameters.~
- [ ] ~Support for binding non-file form parameters (`[FromForm("foo")] string value`)?~ Removed due to https://github.com/dotnet/aspnetcore/issues/34303#issuecomment-894816836
- [x] ~API explorer support for OpenAPI.~
- [ ] Also support `IEnumerable<IFormFile>` for parity with MVC? (See 49c653ed0baf829cb91bb51bb575306c9f1ec0fa)

Relates to #34303 and #35100.

/cc @davidfowl @jchannon

martincostello commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -467,58 +489,113 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+                    }
+                    else
+                    {
+                        try
+                        {
+                            contentValue = await httpContext.Request.ReadFormAsync();
+                        }
+                        catch (IOException ex)
+                        {
+                            Log.RequestFormIOException(httpContext, ex);
+                            return;
+                        }
+                        catch (InvalidDataException ex)
+                        {
+                            Log.RequestFormInvalidDataException(httpContext, ex);
+                            httpContext.Response.StatusCode = 400;
+                            return;
+                        }
```
This is basically copy-pasta from the above block, with more specific log messages. Open to alternatives to make this more DRY and if it doesn't need the more-specific logging.

davidfowl replied:
This should be another lambda as you can't have both a JSON and form body.

martincostello replied:
587d22f605194c763b80a81e06357fa46e6ed310

martincostello commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -582,6 +659,50 @@ private static Expression BindParameterFromValue(ParameterInfo parameter, Expres
                         valueExpression,
                         Expression.Convert(Expression.Constant(parameter.DefaultValue), parameter.ParameterType)));
             }
+            else if (parameter.ParameterType == typeof(IFormFile))
```
Again, copy-pasta from the above block for `typeof(string)`, might be redundant with the change I made to `GetValueFromProperty()` to allow the return type to be different, or the implementation might want to be something completely different.

martincostello replied:
Yep - updated in a9ac8a09ceb3562e68ba424760a408f05596b836.

martincostello commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -986,6 +1112,18 @@ public static void RequiredParameterNotProvided(HttpContext httpContext, string
                 EventName = "RequiredParameterNotProvided")]
             private static partial void RequiredParameterNotProvided(ILogger logger, string parameterType, string parameterName);
 
+            public static void RequestFormIOException(HttpContext httpContext, IOException exception)
+                => RequestFormIOException(GetLogger(httpContext), exception);
+
+            [LoggerMessage(5, LogLevel.Debug, "Reading the request form failed with an IOException.", EventName = "RequestFormIOException")]
+            private static partial void RequestFormIOException(ILogger logger, IOException exception);
+
+            public static void RequestFormInvalidDataException(HttpContext httpContext, InvalidDataException exception)
+                => RequestFormInvalidDataException(GetLogger(httpContext), exception);
+
+            [LoggerMessage(6, LogLevel.Debug, "Reading the request form failed with an InvalidDataException.", EventName = "RequestFormInvalidDataException")]
+            private static partial void RequestFormInvalidDataException(ILogger logger, InvalidDataException exception);
+
```
Might be overkill and could just re-use the body logging.

martincostello commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -467,58 +484,113 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+                    {
+                        try
+                        {
+                            contentValue = await httpContext.Request.ReadFormAsync();
```
This will throw `InvalidOperationException` if the content-type is incorrect. Should that be caught and an HTTP 415 status returned?

https://github.com/dotnet/aspnetcore/blob/8b30d862de6c9146f466061d51aa3f1414ee2337/src/Http/Http/src/Features/FormFeature.cs#L148-L151

martincostello replied:
9d71e8a483fe43c3620bac119b7906b670ee8c9a

martincostello commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -707,6 +779,9 @@ private static Expression BindParameterFromRouteValueOrQueryString(ParameterInfo
             return BindParameterFromValue(parameter, Expression.Coalesce(routeValue, queryValue), factoryContext);
         }
 
+        private static Expression BindParameterFromFormFile(ParameterInfo parameter, MemberExpression property, string key, FactoryContext factoryContext) =>
+            BindParameterFromValue(parameter, GetValueFromProperty(property, key, typeof(IFormFile)), factoryContext);
```
Should this method and `BindParameterFromBody` throw if a parameter for the other is specified?

https://github.com/dotnet/aspnetcore/blob/4c8052fcc1f5ccc1c315721926ed96b553a3f7a6/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L785-L790

i.e. an action with the following signature would be invalid because it can't be both:

```csharp
app.MapPost("/blah", async ([FromBody] MyObject myJson, IFormFile myFile) =>
{
    // Whatever
});
```

davidfowl replied:
Yes it should. This wouldn't make any sense unless we natively supported binding multipart form content (we don't even support this in MVC).

martincostello replied:
8aa7a7dc84f4f715fc05eeabe88e8e9ee553dfb3

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -538,7 +583,7 @@ private static Expression BindParameterFromValue(ParameterInfo parameter, Expres
             var argument = Expression.Variable(parameter.ParameterType, $"{parameter.Name}_local");
 
-            if (parameter.ParameterType == typeof(string))
+            if (parameter.ParameterType == typeof(string) || parameter.ParameterType == typeof(IFormFile))
```
This doesn't seem right. What's this for?

martincostello replied:
Without this the special case for `typeof(string)` was skipped, and then it would throw an `InvalidOperationException` due to `IFormFile` not having a `TryParseMethod()`.

This was just enough to get the tests passing re-using `BindParameterFromValue()`, I wasn't sure what the best way to have it handle the parameter was so I just went with this pending feedback on what the preferred way would be.

I guess a custom method similar to `BindParameterFromService()`?

martincostello replied:
1663381e146e825cea31e4ec86e6bdab06d508d6

davidfowl commented:
Needs support in the [EndpointMetadataApiDescriptionProvider](https://github.com/dotnet/aspnetcore/blob/1795d691005181241617007399abd3c8d10ef32a/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L154) so it shows up property in API explorer. It also needs the [right request content type](https://github.com/dotnet/aspnetcore/blob/1795d691005181241617007399abd3c8d10ef32a/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L220-L226)

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -752,8 +767,64 @@ private static Expression BindParameterFromRouteValueOrQueryString(ParameterInfo
 
-        private static Expression BindParameterFromFormFile(ParameterInfo parameter, MemberExpression property, string key, FactoryContext factoryContext) =>
-            BindParameterFromValue(parameter, GetValueFromProperty(property, key, typeof(IFormFile)), factoryContext);
+        private static Expression BindParameterFromFormFile(ParameterInfo parameter, string key, FactoryContext factoryContext)
```
I see what you mean, it is 90% shared...

martincostello replied:
Once #35167 is merged and I fix-up any conflicts, I'll try factoring it out into a separate method that `BindParameterFromValue()` and `BindParameterFromFormFile()` can both use.

martincostello replied:
be491e0486e82e60eaebf559f48e31686f37c245

pranavkm reviewed (commented):
Allowing form parameters without antiforgery leaves this open to CSRF out of the box. Has that been thought thru?

martincostello commented:
It was mentioned by @davidfowl here https://github.com/dotnet/aspnetcore/issues/34303#issuecomment-894816836 (I assumed _forks_ was a typo for _forms_), but I haven't addressed there here other than removing the `IFormCollection` support that was in the initial commit.

Off the top of my head, if that needs adding in here would that "just" be the case of updating the method that is reading the form to do something similar to what the `ValidateAntiforgeryTokenAuthorizationFilter` class does?

davidfowl commented:
This change looks really good @martincostello!

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -709,13 +727,32 @@ private static Expression BindParameterFromRouteValueOrQueryString(ParameterInfo
 
+        private static Expression BindParameterFromFormFile(ParameterInfo parameter, string key, FactoryContext factoryContext)
+        {
+            if (factoryContext.JsonRequestBodyType is not null)
```
I would add a computed property on the factoryContext `HasBody => JsonRequestBodyType is not null || ReadForm` and use that here and in the JSON logic

martincostello replied:
The problem with doing exactly that is it wouldn't allow you to have more than one `IFormFile` parameter as trying to bind the second one would find the first and then throw.

martincostello commented:
Going to mark this ready for review as I can't make any further progress on this without input from the team about CSRF and the new metadata API suggestion. I also have to keep rebasing it 😄

I guess possible outcomes include:

1. Can't implement as-is due to CSRF concerns
2. Candidate for 6.0 if CSRF concerns and metadata addition(s) approved soon (seems tight to make it to RC1 now?)
3. Gets punted back to 7.0 and CSRF and metadata dealt with when the team has time to come back to it.

davidfowl commented:
@martincostello this is something we're willing to take for .NET 7 and with the latest changes, it'll be possible to simplify this change.

martincostello commented:
@davidfowl - Ah cool - I stopped rebasing as I saw the team was still putting lots of fit and finish on things so was stepping back waiting for the dust to settle. If/when the majority of the changes you have planned are in for 6.0, let me know and I'll rebase this 👍

davidfowl commented:
@martincostello can you revamp this PR?

@rafikiassumani-msft can you assign somebody to review this one?

rafikiassumani-msft commented:
assigned to @halter73

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -672,52 +806,10 @@ private static Expression BindParameterFromValue(ParameterInfo parameter, Expres
             var sourceConstant = Expression.Constant(source);
 
-            if (parameter.ParameterType == typeof(string))
+            if (parameter.ParameterType == typeof(string) ||
+                parameter.ParameterType == typeof(IFormFileCollection))
```
```suggestion
            if (parameter.ParameterType == typeof(string))
```

BindParameterFromValue isn't called for IFormFileCollections, right?

martincostello replied:
I'll double-check before I push up a rebase, but (at least initially when the PR was newish) this was to be able to re-use as much of the code here as I could to reduce duplication, so it was going in here to bind the value.

martincostello replied:
You're right - this is redundant now.

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -643,12 +677,112 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+                    var feature = httpContext.Features.Get<IHttpRequestBodyDetectionFeature>();
+                    if (feature?.CanHaveBody == true)
+                    {
+                        if (!httpContext.Request.HasJsonContentType())
```
```suggestion
                        if (!httpContext.Request.HasFormContentType)
```

Can we add a test case for minimal APIs accepting both form parameters and `BindAsync` parameters?

There's also a lot of duplication between the `if (factoryContext.ParameterBinders.Count > 0)` branch and the `else` branch. Is there a way we can reduce this duplication both here and in HandleRequestBodyAndCompileRequestDelegateForJson? And/or possibly reduce duplication between HandleRequestBodyAndCompileRequestDelegateForJson and HandleRequestBodyAndCompileRequestDelegateForForm?

martincostello replied:
Sure - I erred away from refactoring away the duplication to start with as the way the Lambdas were structured I assumed there was a perf reason it was the way it was.

It was less duplicative initially as there was only a need for an `else`, but some changes between when I opened this and now meant the 2 branches become 4 on a previous rebase.

Unless there's any specific characteristics of why this bit is the way it is, I'll try and factor it out to reduce the duplication.

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -889,6 +1041,46 @@ private static Expression BindParameterFromBindAsync(ParameterInfo parameter, Fa
+            FactoryContext factoryContext,
+            string trackedParameterSource)
+        {
+            factoryContext.FormRequestBodyParameter = parameter;
```
What if there are multiple IFormFile parameters? Does this just clobber the FormRequestBodyParameter?

martincostello replied:
This _should_ handle multiple parameters (there's a test or two for that), but will double-check when I rebase this.

martincostello replied:
You're right - I've refactored this to prefix with `First` and then stopped it being clobbered. It's only used to report a parameter name in a log message if reading the form fails.

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -1291,19 +1494,47 @@ public static void ImplicitBodyNotProvided(HttpContext httpContext, string param
+                UnexpectedFormContentType(GetLogger(httpContext), contentType ?? "(none)");
+            }
+
+            [LoggerMessage(7, LogLevel.Debug, UnexpectedFormContentTypeLogMessage, EventName = "UnexpectedContentType")]
```
```suggestion
            [LoggerMessage(7, LogLevel.Debug, UnexpectedFormContentTypeLogMessage, EventName = "UnexpectedNonFormContentType")]
```

We should probably give this a unique name since we gave it a unique id. Feel free to pick a better name.

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -1380,5 +1611,21 @@ private static string BuildErrorMessageForInferredBodyParameter(FactoryContext f
+            foreach (var kv in factoryContext.TrackedParameters)
+            {
+                errorMessage.AppendLine(FormattableString.Invariant($"{kv.Key,-19} | {kv.Value,-15}"));
+            }
```
Nit: Can we share this with logic with BuildErrorMessageForInferredBodyParameter?

halter73 commented on src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:
```diff
@@ -239,12 +249,17 @@ private static ParameterDescriptor CreateParameterDescriptor(ParameterInfo param
                      parameter.ParameterType == typeof(HttpResponse) ||
                      parameter.ParameterType == typeof(ClaimsPrincipal) ||
                      parameter.ParameterType == typeof(CancellationToken) ||
+                     parameter.ParameterType == typeof(IFormFileCollection) ||
```
Why is this not also `BindingSource.FormFile`?

martincostello replied:
So it'll fall into this block as it binds to a `FormFile`, rather than `Services`.

https://github.com/dotnet/aspnetcore/blob/05cbaa0f4bb99a8d4f54daba4fb7e85483d324aa/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L283-L286

martincostello replied:
Ah, sorry, I see what you mean now.

Yeah I wasn't sure if this counted as `Form` (because that's where it ultimately comes from), `FormFile` (even though there might be multiple files, so then is it being _a_ form file correct?) or `Services`.

I went with `Services` to start with, but I can change it to one of the other two.

halter73 commented on src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:
```diff
@@ -114,17 +114,27 @@ private ApiDescription CreateApiDescription(RouteEndpoint routeEndpoint, string
+                var alreadyHasAnyBodyParameter = apiDescription.ParameterDescriptions.Any(x =>
+                    x.Source == BindingSource.Body ||
+                    x.Source == BindingSource.Form ||
+                    x.Source == BindingSource.FormFile);
```
Why didn't we need this check before and what makes it necessary now?

martincostello replied:
I believe this is just something that wasn't catered for initially as with form files not being supported the code wouldn't get this far with something the parameter description would be wrong for. Without this check, it'll generate a parameter description that doesn't render appropriately in Swagger UI for dealing with a file.

I'll re-check this when I rebase and grab some screenshots of the before-after UI behaviour from rendering the OpenAPI document.

martincostello replied:
Right, so this is because the code currently assumes that if the endpoint has an `IAcceptsMetadata`, it comes from the body and adds the parameter to the parameter descriptions because it's skipped here:

https://github.com/dotnet/aspnetcore/blob/a5a1e9a9d8631caf39366c11511c0cd126a02b34/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L150-L155

This is because previously `IAcceptsMetadata` was only added for a JSON request body:

https://github.com/dotnet/aspnetcore/blob/a5a1e9a9d8631caf39366c11511c0cd126a02b34/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L892-L912

This new check prevents this "default" parameter being added for the body, because a more specific parameter for the `IFormFile` will have been added already. The route already has some `IAcceptsMetadata` in the case of form files as `IAcceptsMetadata` is added for them in these two places:

https://github.com/dotnet/aspnetcore/blob/05cbaa0f4bb99a8d4f54daba4fb7e85483d324aa/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L1044-L1060

https://github.com/dotnet/aspnetcore/blob/05cbaa0f4bb99a8d4f54daba4fb7e85483d324aa/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L1062-L1082

Alternatively I could remove this "default" behaviour for the body by removing the skip for `BindingSource.Body` and having it dealt with explicitly in `CreateApiParameterDescription()` instead.

halter73 replied:
We want to support people manually consuming the request body and then specifying a type for the body "parameter" via `.Accepts<TRequest>(...)` or `[Consumes(...)]`. This is the reason for the current logic that adds an ApiParameterDescription for endpoint IAcceptsMetadata. As a consequence, we skip `source == BindingSource.Body` parameters in CreateApiParameterDescription() just to avoid duplicates.

> Alternatively I could remove this "default" behaviour for the body by removing the skip for BindingSource.Body and having it dealt with explicitly in CreateApiParameterDescription() instead.

I think this would probably the best approach if we could distinguish IAcceptsMetadata added by RequestDelegateFactory by default vs. IAcceptsMetadata added manually. Then we could only add ApiParameterDescriptions for manually-added IAcceptsMetadata. I'm not sure how we'd do that though.

The more consistent approach would be to also skip `source == BindingSource.FormFile`, but this loses info like the name of the form field(s) and whether there are multiple parameters for multiple fields.

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -643,12 +677,112 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+                        }
+                        try
+                        {
+                            formValue = await httpContext.Request.ReadFormAsync();
```
Because we don't have an anti-forgery story for minimal APIs yet, we want to throw a `BadHttpRequestException` if there's a "Cookie" or "Authorization" request header. Or if there's a client certificate. @blowdart 

I'll file a follow up issue to add anti-forgery support when we merge this PR.

martincostello committed:
Support binding to form parameters in minimal actions

Add support for binding to parameters of types IFormFileCollection
and IFormFile on the HttpRequest in minimal actions.

martincostello committed:
Remove dead branch

Method is never called for IFormFileCollection.

martincostello committed:
Add test case for two IFormFile values

Add an explicit test case for handling multiple files.

martincostello committed:
Fix IFormFile with bound parameters

Fix copy-pasta which broke form binding when using BindAsync() and add test.

martincostello committed:
Deduplicate JSON/form body reading

Deduplicate the methods for reading the body as JSON or as a form for with and without bound parameters.

martincostello committed:
Add more invalid parameter tests

Add more tests for invalid body parameter combinations with multiple files.

martincostello committed:
Reduce arguments on TryRead*Async methods

Just accept the HttpContext and the FactoryContext and do the work with them in the static local methods, rather than in the caller.

martincostello committed:
Clarify property name

Clarify that the parameter is just the first body parameter and is used for error reporting. Do not overwrite the first value found if there are multiple values.

martincostello committed:
Rename log message

Make the error case for an invalid form have a unique name compared to a JSON body.

martincostello committed:
Reuse formatting code for parameters

Add a helper method to format the tracked parameters.

martincostello committed:
Disallow form use with authentication

Prevent the form being read if the request is considered to contain a secure value while there are no anti-forgery protections.

martincostello committed:
Rename metadata to IFromFormMetadata

Rename IFromFileMetadata to IFromFormMetadata.

martincostello committed:
Rename test class

Rename test implementation to FromFormAttribute.

martincostello committed:
Support IFromFormMetadata for IFormFileCollection

Support the use of IFromFormMetadata for parameters of type IFormFileCollection. Throw if a name is used for the metadata.

martincostello committed:
Add IFromFormMetadata

Add IFromFormMetadata as an implemented interface.

martincostello committed:
Fix test compilation

Fix tests that wouldn't compile after the rename in 9a0e5e22d51098a4288090d77c2a26b3def28727 that was done from another solution file.
Add TODO for how to correctly handle IFormFileCollection with IFromFormMetadata.

martincostello commented:
I've applied the feedback from https://github.com/dotnet/aspnetcore/issues/35175#issuecomment-963475874, there's just a few TODOs regarding what binding sources to use for which types I'm unsure of which is best.

halter73 commented on src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:
```diff
@@ -239,12 +249,19 @@ private static ParameterDescriptor CreateParameterDescriptor(ParameterInfo param
+        else if (attributes.OfType<IFromFormMetadata>().FirstOrDefault() is { } fromFormAttribute)
+        {
+            // TODO Should this be Form or FormFile if parameter.ParameterType is IFormFileCollection?
+            // Currently it's Services (see below) - which is the more appropriate?
```
I think we should probably use `BindingSource.FormFile` for both `IFormFile` and `IFormFileCollection`. 

@domaindrivendev @RicoSuter @captainsafia

pranavkm replied:
https://github.com/dotnet/aspnetcore/blob/a450cb69b5e4549f5515cdb057a68771f56cefd7/src/Mvc/Mvc.Core/src/Infrastructure/MvcCoreMvcOptionsSetup.cs#L131-L134

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -611,44 +625,204 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+            Debug.Assert(factoryContext.JsonRequestBodyParameter is not null, "TryReadFormAsync() should not be called if there is no JSON body parameter.");
+
+            var bodyType = factoryContext.JsonRequestBodyParameter.ParameterType;
+            var parameterTypeName = TypeNameHelper.GetTypeDisplayName(factoryContext.JsonRequestBodyParameter.ParameterType, fullName: false);
```
Anything remotely expensive should be moved out of the per-request logic. We should be capturing the type name. I think this is also a good opportunity to capture only what we need from the FactoryContext.

martincostello replied:
Yeah I did spot this one, and at one point I had this in the catch block as that was the only place the value was used. I moved it back as it was already like this, but meant to ask if it was intentional.

Will move it so it's only done if needed.

halter73 replied:
Previously, we always captured the `parameterTypeName` during the call to `RequestDelegateFactory.Create(...)` rather than lazily doing the reflection every time there's an error during the request.

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -835,8 +966,68 @@ private static Expression BindParameterFromValue(ParameterInfo parameter, Expres
+                Expression.Convert(Expression.Constant(parameter.DefaultValue), parameter.ParameterType)));
+    }
+
+    private static Expression BindParameterFromProperty(ParameterInfo parameter, MemberExpression property, string key, FactoryContext factoryContext, string source, Type? returnType = null) =>
```
Nit: This is no longer called with a non-default `returnType` right?

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -239,6 +249,21 @@ private static Expression CreateArgument(ParameterInfo parameter, FactoryContext
+                if (!string.IsNullOrEmpty(formAttribute.Name))
+                {
+                    throw new NotSupportedException(
+                        $"Assigning a value to the {nameof(IFromFormMetadata)}.{nameof(IFromFormMetadata.Name)} property is not supported for parameters of type {nameof(IFormFileCollection)}.");
```
Do we throw any startup errors when this attribute is applied to parameters that aren't either an `IFormFile` or `IFormFileCollection`?

martincostello replied:
I'll check - assuming not, shall I add a catch-all for any other type than the two new ones?

halter73 replied:
Sounds good!

martincostello replied:
The current behaviour is the below, but I've added a special case that throws `NotSupportedException` instead.

```
typeof(System.ArgumentException): Expression of type 'Microsoft.AspNetCore.Http.IFormFile' cannot be used for assignment to type 'System.Boolean'
```

halter73 commented on src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:
```diff
@@ -114,17 +114,27 @@ private ApiDescription CreateApiDescription(RouteEndpoint routeEndpoint, string
-            var parameterDescription = new ApiParameterDescription
+            // Do not add an implicit body parameter description if a parameter already
+            // exists for something derived from the body's content, such as a form file.
+            var alreadyHasAnyBodyParameter = apiDescription.ParameterDescriptions.Any(x =>
+                x.Source == BindingSource.Body ||
+                x.Source == BindingSource.Form ||
+                x.Source == BindingSource.FormFile);
```
> Alternatively I could remove this "default" behaviour for the body by removing the skip for BindingSource.Body and having it dealt with explicitly in CreateApiParameterDescription() instead.

Let's do this for consistency. The main thing that we want to have keep working is if there is no `BindingSource.Body` or `BindingSource.FormFile` parameter by "default" and someone added IAcceptsMetadata explicitly with `.Accepts<TRequest>(...)` or `[Consumes(...)]` indicating they're manually reading the body, we should continue to create this "fake" ApiParameterDescription from IAcceptsMetadata .

If we know we're consuming the body on behalf of the app, we can just assume that the IAcceptsMetadata is what we added by default. If someone tries to add custom IAcceptsMetadata despite that, it's certainly wrong, so no big loss.

Also nit:

```suggestion
            var alreadyHasAnyBodyParameter = apiDescription.ParameterDescriptions.Any(x =>
                x.Source == BindingSource.Body ||
                x.Source == BindingSource.FormFile);
```

I don't see the point in looking for `BindingSources` we'll never create at this point.

martincostello replied:
Fixed nit.

Still thinking about the best way to tackle this, as I've tried a few different ways to remove the default/skip body parameter but they all seemed to break _something_ in the tests.

halter73 replied:
What tests break? I'm okay with changing some of the tests if they're asserting something nonsensical like custom IAcceptsMetadata being used to generate an ApiParameterDescription when we're consuming an application/json request body on behalf of the app. After this change, custom IAcceptsMetadata won't be used to generate an ApiParameterDescription when we're consuming a multipart/form-data on behalf of the app, so at least this is consistent.

martincostello replied:
There were some tests checking there was no body parameter at all, where there was now a default. There were others that would fail because there was no implicit accept metadata where it was expected because the unit tests don't run through the real RequestDelegateFactory code because it's all mocked routing.

For the latter I was torn between changing it to do so to make it more realistic vs. the fact the two implementation details were becoming aware of each other.

I'll come back to it with fresh eyes tomorrow and maybe it'll all click how best to refactor it.

martincostello committed:
Capture values on start-up

Revert changes from 599e72c17741089a40c18e0da36659e59fa11d58.

martincostello committed:
Remove unused parameter

Remove unused optional parameter.

martincostello committed:
Throw if [FromForm] used with invalid types

Throw a NotSupportedException if [FromForm]/IFromFormMetadata is used for parameters of types other than IFormFile and IFormFileCollection.

martincostello committed:
Update IFormFileCollection BindingSource

Update IFormFileCollection to bind as BindingSource.FormFile.

martincostello committed:
Remove BindingSource.Form case

Remove check for BindingSource.Form as that's not a used value with Minimal APIs.

martincostello committed:
Merge branch 'main' into Minimal-APIs-IForm-Support-34303

martincostello committed:
Update default body parameter handling

Update the default parameter handling for inferred body parameters so the default parameter is only added via IAcceptsMetadata if there were no other parameters already defined.

martincostello committed:
Fix typo

Add missing >.

halter73 commented:
Thanks @martincostello!! 🎉

martincostello commented:
Thanks @halter73 - if you let me know when the agreed CSRF support lands in main, I'd be happy to do a follow-up PR to take out the guards for the authorized requests added in 45b30d2efe3f993a0b25326d834e4406a300cf44.
